src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -231,4 +225,95 @@ protected void performCleanup() {
             // no action
         }
     }
+
+    /**
+     * A specialized implementation that tracks phantom cleanables.
+     * Backing storage is expanded and trimmed automatically.
+     * Insert/remove run in amortized constant time.
+     */
+    static final class PhantomCleanableList {
+        private static final int MIN_CAPACITY = 16;
+        private final Object lock = new Object();
```

Comment:
Why use another lock object when identity of this list is not escaped elsewhere?

Reply:
Mostly hygiene for not exposing the lock to external classes. But yeah, we can simplify by just doing `synchronized`. I'll push the update soon.